File: json/src/main/java/top/andnux/json/JsonRequestBodyConverter.java
package top.andnux.json;/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import com.alibaba.fastjson.JSON;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import retrofit2.Converter;

import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.charset.Charset;

final class JsonRequestBodyConverter<T> implements Converter<T, RequestBody> {

    private static final MediaType MEDIA_TYPE = MediaType.get("application/json; charset=UTF-8");
    private static final Charset UTF_8 = Charset.forName("UTF-8");
    private Type mType;

    public JsonRequestBodyConverter(Type mType) {
        this.mType = mType;
    }

    @Override
    public RequestBody convert(T value) throws IOException {
        String json = JSON.toJSONString(value);
        return RequestBody.create(MEDIA_TYPE, json.getBytes(UTF_8));
    }
}

File: net/src/main/java/top/andnux/net/retrofit/converter/JsonRequestBodyConverter.kt
package top.andnux.net.retrofit.converter

/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import java.io.IOException
import java.lang.reflect.Type
import java.nio.charset.Charset

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody
import retrofit2.Converter
import top.andnux.json.JsonProxy

internal class JsonRequestBodyConverter<T>(private val mJsonAdapter: JsonProxy, private val mType: Type) : Converter<T, RequestBody> {

    @Throws(IOException::class)
    override fun convert(value: T): RequestBody {
        val json = mJsonAdapter.toJSONString(value)
        return RequestBody.create(MEDIA_TYPE, json.toByteArray(UTF_8))
    }

    companion object {
        private val MEDIA_TYPE = "application/json; charset=UTF-8".toMediaType()
        private val UTF_8 = Charset.forName("UTF-8")
    }
}
